be/src/common/statusor.h

```diff
@@ -702,4 +703,29 @@ void StatusOr<T>::IgnoreError() const {
     // no-op
 }
 
+#define ASSIGN_OR_RETURN_IMPL(varname, lhs, rhs) \
+    auto&& varname = (rhs);                      \
+    RETURN_IF_ERROR(varname);                    \
+    lhs = std::move(varname).value();
+
+// ASSIGN_OR_RETURN is modelled after Apache Arrow's ARROW_ASSIGN_OR_RAISE macro.
+//
+// Execute an expression that returns a StatusOr, extracting its value
+/// into the variable defined by `lhs` (or returning a Status on error).
+//
+// Example: Assigning to a new value:
+//   ARROW_ASSIGN_OR_RETURN(auto value, MaybeGetValue(arg));
+//
+// Example: Assigning to an existing value:
+//   ValueType value;
+//   ARROW_ASSIGN_OR_RETURN(value, MaybeGetValue(arg));
+//
+// WARNING: ARROW_ASSIGN_OR_RETURN expands into multiple statements;
+// it cannot be used in a single statement (e.g. as the body of an if
+// statement without {})!
+//
+// WARNING: ARROW_ASSIGN_OR_RETURN `std::move`s its right operand. If you have
+// an lvalue StatusOr which you *don't* want to move out of cast appropriately.
+#define ASSIGN_OR_RETURN(lhs, rhs) ASSIGN_OR_RETURN_IMPL(VARNAME_LINENUM(value_or_err), lhs, rhs)
```

Comment:
Can you use some hack to avoid misusing, assign rhs with a lvalue?

Reply:
`rhs` has to be an expression of type `StatusOr`, assign `rhs` with a lvalue of different type will generate a compiler error

Reply:
When rhs is a lvalue of StarusOr, this macro will change it. Caller may not know this and it is not his/her intent.
How to avoid using it like this.

Reply:
> 
 The callers should know this behavior, as mentioned in the above comment. It's the caller's duty to follow a utility tool's documentation.

Reply:
Passing an lvalue as the `rhs` is not an error but a well-defined behavior, but the caller should know the side-effect, which has been mentioned in the comment of the macro.

Reply:
Can you use same magic to avoid this behavior in compiling time?

Reply:
> Can you use same magic to avoid this behavior in compiling time?

Sorry for the late replay, enable pass an lvalue as `rhs` is by-design